beacon-chain/sync/rpc_beacon_blocks_by_range.go

```diff
@@ -145,24 +145,27 @@ func (r *Service) writeBlockRangeToStream(ctx context.Context, startSlot, endSlo
 		roots = append([][32]byte{genRoot}, roots...)
 	}
 	blks, roots = r.sortBlocksAndRoots(blks, roots)
-	checkpoint, err := r.db.FinalizedCheckpoint(ctx)
-	if err != nil {
-		log.WithError(err).Error("Failed to retrieve finalized checkpoint")
-		r.writeErrorResponseToStream(responseCodeServerError, genericError, stream)
-		traceutil.AnnotateError(span, err)
-		return err
-	}
 	for i, b := range blks {
 		if b == nil || b.Block == nil {
 			continue
 		}
 		blk := b.Block
 
+		// check that the block is valid according to the request
+		// and part of the canonical chain.
 		isRequestedSlotStep := (blk.Slot-startSlot)%step == 0
-		isRecentUnfinalizedSlot := blk.Slot >= helpers.StartSlot(checkpoint.Epoch+1) || checkpoint.Epoch == 0
-		if isRequestedSlotStep && (isRecentUnfinalizedSlot || r.db.IsFinalizedBlock(ctx, roots[i])) {
+		isCanonical, err := r.chain.IsCanonical(ctx, roots[i])
```

Comment:
We don't need to check if a block is canonical if it is not on the requested slot step. Imagine the step was 512, we would be running IsCanonical 512 times for 2 blocks. 

Please refactor to only call this method when `isRequestedSlotStep == true`

Reply:
I will fix this.